uefi-raw/src/protocol/network/tcpv4.rs

```diff
@@ -0,0 +1,513 @@
+    /// asynchronous operations are signaled, and any buffers used for
+    /// TCP network traffic are flushed.
+    pub close:
+        unsafe extern "efiapi" fn(this: *mut Self, close_token: *mut Tcpv4CloseToken) -> Status,
+
+    /// Abort an asynchronous connection, listen, transmission or
+    /// receive request.
+    ///
+    /// The `cancel` function aborts a pending connection, listen,
+    /// transmit or receive request. If `completion_token` is not
+    /// `NULL` and the token is in the connection, listen,
+    /// transmission or receive queue when it is being cancelled, its
+    /// `status` will be set to [`Status::ABORTED`] and then `event`
+    /// will be signaled. If the token is not in one of the queues,
+    /// which usually means that the asynchronous operation has
+    /// completed, [`Status::NOT_FOUND`] is returned. If
+    /// `completion_token` is `NULL`, all asynchronous tokens issued
+    /// by [`Tcpv4Protocol::connect`], [`Tcpv4Protocol::accept`],
+    /// [`Tcpv4Protocol::transmit`] and [`Tcpv4Protocol::receive`]
+    /// will be aborted.
+    pub cancel: unsafe extern "efiapi" fn(
+        this: *mut Self,
+        completion_token: *mut Tcpv4CompletionToken,
+    ) -> Status,
+
+    /// Poll to receive incoming data and transmit outgoing segments.
+    ///
+    /// The `poll` function increases the rate that data is moved
+    /// between the network and application and can be called when the
+    /// TCP instance is created successfully. Its use is optional.
+    ///
+    /// In some implementations, the periodical timer in the MNP
+    /// driver may not poll the underlying communications device fast
+    /// enough to avoid dropping packets. Drivers and applications
+    /// that are experiencing packet loss should try calling the
+    /// `poll` function at a high frequency.
+    pub poll: unsafe extern "efiapi" fn(this: *mut Self) -> Status,
+}
+
+impl Tcpv4Protocol {
+    /// The GUID for the TCPv4 protocol.
+    ///
+    /// Defined in the [UEFI Specification, Section 28.1.4](https://uefi.org/specs/UEFI/2.11/28_Network_Protocols_TCP_IP_and_Configuration.html#efi-tcp4-protocol).
+    pub const GUID: Guid = guid!("65530BC7-A359-410F-B010-5AADC7EC2B62");
+
+    /// The GUID for the TCPv4 service binding protocol.
+    ///
+    /// Defined in the [UEFI Specification, Section 28.1.2](https://uefi.org/specs/UEFI/2.11/28_Network_Protocols_TCP_IP_and_Configuration.html#efi-tcp4-service-binding-protocol).
+    pub const SERVICE_BINDING_GUID: Guid = guid!("00720665-67EB-4a99-BAF7-D3C33A1C7CC9");
+}
+
+newtype_enum! {
+    pub enum Tcpv4ConnectionState: i32 => {
+        CLOSED = 0,
+        LISTEN = 1,
+        SYN_SENT = 2,
+        SYN_RECEIVED = 3,
+        ESTABLISHED = 4,
+        FIN_WAIT1 = 5,
+        FIN_WAIT2 = 6,
+        CLOSING = 7,
+        TIME_WAIT = 8,
+        CLOSE_WAIT = 9,
+        LAST_ACK = 10,
+    }
+}
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Ipv4ModeData {
```

Comment:
I think this type (and some of the following ones) belong in `uefi-raw/src/protocol/network/ip4.rs`.

Reply:
fixed